### PR TITLE
buildGoModule: add vendorHash

### PR DESCRIPTION
**Pull request by ajs124**
###### Description of changes
the _unset hack is kind of ugly, but it needs to default to something
and it can't be null, because that already has special meaning

###### Things done
- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/go-modules/generic/default.nix [hunk 1]
```diff
@@ -55,11 +57,24 @@
 with builtins;
 
 assert goPackagePath != "" -> throw "`goPackagePath` is not needed with `buildGoModule`";
+assert (vendorSha256 == "_unset" && vendorHash == "_unset") -> throw "either `vendorSha256` or `vendorHash` is required";
+assert vendorSha256 != null && vendorSha256 != "_unset" && vendorHash != null && vendorHash != "_unset"
```
Comment (ajs124) [inline pkgs/development/go-modules/generic/default.nix:61]: I think this still allows for both to be set and `null`, if I'm not mistaken.

File: pkgs/development/go-modules/generic/default.nix [hunk 2]
```diff
@@ -55,11 +57,24 @@
 with builtins;
 
 assert goPackagePath != "" -> throw "`goPackagePath` is not needed with `buildGoModule`";
+assert (vendorSha256 == "_unset" && vendorHash == "_unset") -> throw "either `vendorSha256` or `vendorHash` is required";
```
Comment (zowoq) [inline pkgs/development/go-modules/generic/default.nix:60]: Please put `vendorHash` first in the message.

```suggestion
assert (vendorSha256 == "_unset" && vendorHash == "_unset") -> throw "either `vendorSha256` or `vendorHash` is required";
```

File: pkgs/development/go-modules/generic/default.nix [hunk 3]
```diff
@@ -98,7 +113,7 @@ let
       fi
     '' + ''
       if [ -d vendor ]; then
-        echo "vendor folder exists, please set 'vendorSha256 = null;' in your expression"
+        echo "vendor folder exists, please set 'vendorSha256 = null;' or 'vendorHash = null;' in your expression"
```
Comment (zowoq) [inline pkgs/development/go-modules/generic/default.nix:116]: First here as well.

**Review by zowoq — COMMENTED** [review]
Thanks, I'd started working on this a while back but never got around to finishing it.

Can we set a default hash type so `vendorHash = "";` doesn't do this: `error: empty hash requires explicit hash type`?

**Comment by zowoq** [comment]
```diff
diff --git a/pkgs/development/go-modules/generic/default.nix b/pkgs/development/go-modules/generic/default.nix
index 081e736321c..72105d40cf1 100644
--- a/pkgs/development/go-modules/generic/default.nix
+++ b/pkgs/development/go-modules/generic/default.nix
@@ -19,19 +19,19 @@
 # path to go.mod and go.sum directory
 , modRoot ? "./"
 
-# vendorSha256 is the sha256 of the vendored dependencies
+# vendorHash is the SRI hash of the vendored dependencies
 #
-# if vendorSha256 is null, then we won't fetch any dependencies and
+# if vendorHash is null, then we won't fetch any dependencies and
 # rely on the vendor folder within the source.
-, vendorSha256 ? "_unset"
-# same as vendorSha256, but allows for using a SRI hash to specify algorithm
 , vendorHash ? "_unset"
+# same as vendorHash, but allows for using a base32 hash to specify algorithm
+, vendorSha256 ? "_unset"
 # Whether to delete the vendor folder supplied with the source.
 , deleteVendor ? false
 # Whether to fetch (go mod download) and proxy the vendor directory.
 # This is useful if your code depends on c code and go mod tidy does not
 # include the needed sources to build or if any dependency has case-insensitive
-# conflicts which will produce platform dependant `vendorSha256` checksums.
+# conflicts which will produce platform dependant `vendorHash` checksums.
 , proxyVendor ? false

```
```diff
diff --git a/doc/languages-frameworks/go.section.md b/doc/languages-frameworks/go.section.md
index 9c67a514335..8616d64e7c4 100644
--- a/doc/languages-frameworks/go.section.md
+++ b/doc/languages-frameworks/go.section.md
@@ -11,8 +11,8 @@ The function `buildGoModule` builds Go programs managed with Go modules. It buil
 
 In the following is an example expression using `buildGoModule`, the following arguments are of special significance to the function:
 
-- `vendorSha256`: is the hash of the output of the intermediate fetcher derivation. `vendorSha256` can also take `null` as an input. When `null` is used as a value, rather than fetching the dependencies and vendoring them, we use the vendoring included within the source repo. If you'd like to not have to update this field on dependency changes, run `go mod vendor` in your source repo and set `vendorSha256 = null;`
-- `proxyVendor`: Fetches (go mod download) and proxies the vendor directory. This is useful if your code depends on c code and go mod tidy does not include the needed sources to build or if any dependency has case-insensitive conflicts which will produce platform dependant `vendorSha256` checksums.
+- `vendorHash`: is the hash of the output of the intermediate fetcher derivation. `vendorHash` can also take `null` as an input. When `null` is used as a value, rather than fetching the dependencies and vendoring them, we use the vendoring included within the source repo. If you'd like to not have to update this field on dependency changes, run `go mod vendor` in your source repo and set `vendorHash = null;`
+- `proxyVendor`: Fetches (go mod download) and proxies the vendor directory. This is useful if your code depends on c code and go mod tidy does not include the needed sources to build or if any dependency has case-insensitive conflicts which will produce platform dependant `vendorHash` checksums.
 
 ```nix
 pet = buildGoModule rec {
@@ -26,7 +26,7 @@ pet = buildGoModule rec {
     sha256 = "0m2fzpqxk7hrbxsgqplkg7h2p7gv6s1miymv3gvw0cz039skag0s";
   };
 
-  vendorSha256 = "1879j77k96684wi554rkjxydrj8g3hpp0kvxz03sd8dmwr3lh83j";
+  vendorHash = "sha256-ciBIR+a1oaYH+H1PcC8cD8ncfJczk1IiJ8iYNM+R6aA=";
 
   meta = with lib; {
     description = "Simple command-line snippet manager, written in Go";
```

**Comment by zowoq** [comment]
Pushed my diff here: https://github.com/NixOS/nixpkgs/commit/de37cdc99d85a8eaf1b76de071573986c73ed043

**Comment by ajs124** [comment]
> Can we set a default hash type so `vendorHash = "";` doesn't do this: `error: empty hash requires explicit hash type`?

Is something like 45f457a75a3f1d10b3cdff2e01cd263b559262e8 what you had in mind?

File: pkgs/development/go-modules/generic/default.nix (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/development/go-modules/generic/default.nix:62]: We shouldn't allow bot to be set.

File: pkgs/development/go-modules/generic/default.nix [hunk 4]
```diff
@@ -55,11 +58,24 @@
 with builtins;
```
Comment (SuperSandro2000) [inline pkgs/development/go-modules/generic/default.nix:58]:
```suggestion
with lib;
```

**Comment by zowoq** [comment]
> > Can we set a default hash type so `vendorHash = "";` doesn't do this: `error: empty hash requires explicit hash type`?
> 
> Is something like [45f457a](https://github.com/NixOS/nixpkgs/commit/45f457a75a3f1d10b3cdff2e01cd263b559262e8) what you had in mind?

Yes, that works.

File: pkgs/development/go-modules/generic/default.nix (hunk 4 above)
Comment (zowoq) [reply pkgs/development/go-modules/generic/default.nix:58]: Let's keep this PR focused on the vendor changes.

File: pkgs/development/go-modules/generic/default.nix [hunk 5]
```diff
@@ -55,11 +58,24 @@
...
 
 assert goPackagePath != "" -> throw "`goPackagePath` is not needed with `buildGoModule`";
+assert (vendorSha256 == "_unset" && vendorHash == "_unset") -> throw "either `vendorHash` or `vendorHash` is required";
+assert vendorSha256 != null && vendorSha256 != "_unset" && vendorHash != null && vendorHash != "_unset"
```
Comment (zowoq) [inline pkgs/development/go-modules/generic/default.nix:62]: > I think this still allows for both to be set and null, if I'm not mistaken.

Sorry, I missed this the first time. As Sandro says don't we only want to allow one to be set? e.g.

```suggestion
assert (vendorSha256 != "_unset" && vendorHash != "_unset")
```

File: pkgs/development/go-modules/generic/default.nix [hunk 6]
```diff
@@ -55,11 +58,24 @@
...
+  -> throw "both `vendorHash` and `vendoSha256` set. only one can be set.";
```
Comment (zowoq) [inline pkgs/development/go-modules/generic/default.nix:63]: typo
```suggestion
  -> throw "both `vendorHash` and `vendorSha256` set. only one can be set.";
```

**Comment by zowoq** [comment]
Can you squash all the commits except for f822b5c63d7872b6ce4dca6bd100335d05924f98 please?

File: pkgs/development/go-modules/generic/default.nix [hunk 7]
```diff
@@ -55,11 +58,23 @@
 with builtins;
 
 assert goPackagePath != "" -> throw "`goPackagePath` is not needed with `buildGoModule`";
+assert (vendorSha256 == "_unset" && vendorHash == "_unset") -> throw "either `vendorHash` or `vendorHash` is required";
```
Comment (zowoq) [inline pkgs/development/go-modules/generic/default.nix:61]: throw has `vendorHash` twice.

```suggestion
assert (vendorSha256 == "_unset" && vendorHash == "_unset") -> throw "either `vendorHash` or `vendorSha256` is required";
```

Comment (ajs124) [reply pkgs/development/go-modules/generic/default.nix]: good that you're reviewing this. apparently I've only been working on this half-asleep :grimacing:

**Comment by zowoq** [comment]
Thanks!

**Pull request merged**